[speculative-execution] Hoists debug values unnecessarily.

CarlosAlbertoEnciso opened the pull request:
After https://reviews.llvm.org/D81730:
`SpeculativeExecutionPass::considerHoistingFromTo` hoists instructions, including debug intrinsics, as long as none of their used values are instructions that appear prior in the block that are not being hoisted.

This behaviour has been duplicated for DPValues to get rid of a binary difference.

The correct solution is not hoist these debug values at all, whichever format they're in.

CarlosAlbertoEnciso committed:
[speculative-execution] Hoists debug values unnecessarily.

After https://reviews.llvm.org/D81730:
`SpeculativeExecutionPass::considerHoistingFromTo` hoists
instructions, including debug intrinsics, as long as none
of their used values are instructions that appear prior
in the block that are not being hoisted.

This behaviour has been duplicated for DPValues to get
rid of a binary difference.

The correct solution is not hoist these debug values at
all, whichever format they're in.

OCHyams commented on llvm/lib/Transforms/Scalar/SpeculativeExecution.cpp:
```diff
@@ -260,10 +260,31 @@ static InstructionCost ComputeSpeculationCost(const Instruction *I,
+//   call void @llvm.dbg.value(%x, !"x", !DIExpression())
+//   call void foo()
+//
+// If SpeculativeExecution might decide to hoist the 'y * z' calculation
```
nit: remove `If`?

OCHyams reviewed (approved):
LGTM, this change sounds right in isolation. I have a slight worry that this could lead to missing debug records if everything else in the block is hoisted, and then that block (containing only debug records) is deleted later. But ultimately that probably shouldn't be a concern for this pass; passes that delete blocks are responsible for making sure the right thing happens in those cases.

CarlosAlbertoEnciso committed:
[speculative-execution] Hoists debug values unnecessarily.

After https://reviews.llvm.org/D81730:
`SpeculativeExecutionPass::considerHoistingFromTo` hoists
instructions, including debug intrinsics, as long as none
of their used values are instructions that appear prior
in the block that are not being hoisted.

This behaviour has been duplicated for DPValues to get
rid of a binary difference.

The correct solution is not hoist these debug values at
all, whichever format they're in.

nit: Remove 'If' from comment.

CarlosAlbertoEnciso committed:
Merge branch 'main' into speculative-execution-hoists-debug-values
